File: 백준/Gold/1261. 알고스팟/알고스팟.py
from collections import deque

# 미로 nxm 미로는 빈 방 또는 벽
# 이동은 상하좌우
# 벽 부수기 최소 몇개의 벽을 부숴야 하는지 (1,1) -> (n,m)으로 이동
m,n = map(int, input().split())
g = [list(map(int, input())) for _ in range(n)]

INF = int(1e9)
dis = [[INF] * m for _ in range(n)]
dx = [-1,0,1,0]
dy = [0,1,0,-1]
def isInner(x,y):
    if 0<=x<n and 0<=y<m:
        return True
    return False
def BFS(startX,startY):
    dq = deque()
    dq.append((startX,startY,0)) # 현재 위치, 벽 부순 횟수
    while dq:
        x,y,t = dq.popleft() # 현재 위치, 부순 횟수(거리라고 생각)
        if (x,y) == (n-1,m-1): # 도착점 도착
            if t < dis[x][y]: # 갱신 가능
                dis[x][y] = t
            continue
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if not isInner(nx,ny): continue
            if g[nx][ny] == 0: # 빈방 그냥 이동 가능
                if t < dis[nx][ny]: # 현재까지의 횟수가 기존 횟수보다 작으면
                    dis[nx][ny] = t
                    dq.append((nx,ny,t))
            elif g[nx][ny] == 1: # 벽 -> 부숴야
                if t + 1 < dis[nx][ny]: # 현재까지 벽 부순 횟수
                    dis[nx][ny] = t + 1
                    dq.append((nx,ny,t+1))

BFS(0,0)
print(dis[n-1][m-1])

File: 백준/Gold/1261. 알고스팟/알고스팟.java
import javax.xml.crypto.Data;
import java.awt.*;
import java.io.*;
import java.util.*;


public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int n,m,min_cnt;
    static int[][] g;
    static int[][] dis; // ! 벽 최소 제거 횟수.
    static int endX,endY, INF = 1000000;

    private static class Point{
        int x, y, cnt; // 벽을 부순 횟수

        Point(int x, int y, int cnt) {
            this.x = x;
            this.y = y;
            this.cnt = cnt;
        }
    }
    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        g = new int[n][m];
        dis = new int[n][m];
        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < m; j++) {
                g[i][j] = line.charAt(j) - '0';
            }
        }

        for (int i = 0; i < n; i++) {
            Arrays.fill(dis[i], INF);
        }
        min_cnt = INF;
        endX = n-1;
        endY = m-1;

        BFS(0, 0);
        System.out.println(min_cnt);
    }

    static int[] dx = {-1, 0, 1, 0};
    static int[] dy = {0, 1, 0, -1};

    public static boolean isInner(int x, int y) {
        if (x >= 0 && x < n && y >= 0 && y < m) {
            return true;
        }
        return false;
    }

    public static void BFS(int startX, int startY) {
        Deque<Point> dq = new ArrayDeque<>();
        dq.offer(new Point(startX, startY, 0));
        dis[startX][startY] = 0;
        while (!dq.isEmpty()) {
            Point now = dq.poll();
            if (now.x == endX && now.y == endY) {
                if (now.cnt < min_cnt) {
                    min_cnt = now.cnt;
                }
                continue; // ! 최단으로 확인해야 하므로
            }
            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if(!isInner(nx,ny)) continue;
                if (g[nx][ny] == 0) { // 이동 가능
                    if (now.cnt < dis[nx][ny]) {
                        dis[nx][ny] = now.cnt;
                        dq.offer(new Point(nx, ny, now.cnt));
                    }
                } else if (g[nx][ny] == 1) { // ! 벽.
                    if (now.cnt + 1 < dis[nx][ny]) {
                        dis[nx][ny] = now.cnt + 1;
                        dq.offer(new Point(nx, ny, now.cnt + 1));
                    }
                }
            }
        }
    }

}